### ISSUE TITLE
Selectively Enable ExecutionSpace

### ISSUE THREAD

**Issue by flying-tiger** [kokkos/kokkos#2623]
Hello, 

Is it possible to configure Kokkos with device support (e.g. Kokkos_ENABLE_CUDA=ON), but not require a device to be present at runtime? 

Here's my scenario: I would like to build two version of my application, one that runs exclusively on host, one that runs on the device (easy enough using a macro to select an explicit execution space in the program). However, since both versions are built in the same CMake project with the same Kokkos configuration, the host-only version still looks for a CUDA device and terminates when it can't find one. Ideally, I would like an option to disable CUDA support for the host-only version so it can run on a broader set of platforms, without having to do a separate configure/build. Something runtime (e.g. flag to initialize) or compile time (e.g. separate build targets in CMake for kokkos and kokkos_host_only) would work. 

Thanks!

**Comment by mhoemmen** [comment]
It sounds like you are requesting the following feature: Even if an execution space is enabled at configure time, you would like to have the option to disable Kokkos' support for it at run time.  Is that correct?

This is currently not possible.  If you enable CUDA support in Kokkos at configure time, then `Kokkos::initialize` will attempt to start up Kokkos' CUDA support at run time.

**Comment by flying-tiger** [comment]
Ok, thanks. Is this something you think could be done with reasonable effort? I’m willing to contributing the feature. Any preference on the CMake vs. flag to initialize solution?

**Comment by mhoemmen** [comment]
@flying-tiger Other users have requested this feature.  Here is their use case:

  - They are running on a system with N CPU cores and M GPUs per node, with N >> M.
  - Their problems are limited by total available CPU memory.
  - They want to run with N MPI processes (e.g., because MPI parallelization is faster than OpenMP on those CPUs).
  - However, starting the CUDA driver consumes a lot of memory on each MPI process.
  - Thus, they want to start CUDA only on M MPI processes, and use MPI to move data between the remaining N-M "CPU only" processes and the M processes that each manage a GPU.

They could work around this issue by building two versions of the application, one with CUDA enabled and the other with CUDA disabled, and launching different executables on different MPI processes.  (We sometimes call this ["MPMD"](https://en.wikipedia.org/wiki/Flynn%27s_taxonomy#Multiple_programs,_multiple_data_streams_(MPMD)).)  However, that would complicate the build system and testing.

**Comment by mhoemmen** [comment]
@flying-tiger wrote:
> I’m willing to contribute the feature.

We very much appreciate it; thanks! :-)

> Any preference on the CMake vs. flag to initialize solution?

I'm not sure if I understand what you mean.  It sounds like the point is to have a run-time option to disable support for options that were enabled at compile time.  Wouldn't that imply a "flag," not a new CMake option?

> Is this something you think could be done with reasonable effort?

I don't think it's technically difficult.  The issue is that it would complicate Kokkos by adding new error states.  Also, the default execution and memory spaces are set at configure time, so this run-time option would make any use of the default execution or memory spaces a run-time error.  Users would need to use spaces explicitly for things like `View` types and `*Policy` arguments to `parallel_*`.  For example, the following code would cause a run-time error:
```c++
const int N = 42;

// Calls cudaMalloc*; set elements to 0 via CUDA parallel_for
Kokkos::View<double*> x("x", N);

// If you give parallel_* an integer instead of RangePolicy,
// then Kokkos will run on the default execution space.  This is
// Kokkos::Cuda by default if CUDA is enabled at configure time.
Kokkos::parallel_for("my loop", N, KOKKOS_LAMBDA (int k) { x(k) = k+1.0; });
```
You would need to write code like this:
```c++
const int N = 42;

// These could be anything that's not Cuda*.
using execution_space = Kokkos::DefaultHostExecutionSpace;
using memory_space = Kokkos::HostSpace;

using range_policy = Kokkos::RangePolicy<execution_space, int>;
range_policy range(0, N);
// You could still use KOKKOS_LAMBDA here, but it would be
// a waste, since it would unnecessarily mark the lambda as
// __host__ __device__ when you only ever would run on host.
Kokkos::parallel_for("my loop", range, [=] (int k) { x(k) = k+1.0; });
```

**Comment by flying-tiger** [comment]
> I'm not sure if I understand what you mean. It sounds like the point is to have a run-time option to disable support for options that were enabled at compile time.

I see two possible paths here: 

1. Leave CMake is-as, all applications link to Kokkos::kokkos, and introduce runtime flags to selectively bypass initialization of certain backends, e.g. CUDA. As you note, in the case where  bypass the default parallel execution device, it makes the associated default spaces unusable, an user must explicitly specify spaces in their code.

2. Update the CMake build system to introduce Kokkos::kokkos_host_only, a library target that leaves device spaces disabled even when Kokkos_ENABLE_CUDA=ON. Applications then choose which target to link to, depending on need. This allows creating a "MPMD" applications in the same build tree. It also has the upside of ensuring default spaces remain usable. Downside is the selection must be made a compile time (fine for my use case).

Seems like (2) meets my needs and the MPMD use case while sticking a bit closer to the "single source" principal. I'll look into that first unless we think there is a compelling need for true runtime-selectable initialization.

**Comment by mhoemmen** [comment]
@flying-tiger Thanks for explaining!  The Kokkos team will need to discuss this and make a decision.  The other application would actually prefer Approach 1, since it would simplify their build system.  I can see the value of Approach 2, but you can't just build two separate libraries; you would also need to install two separate sets of header files (or at least the header file(s) generated at configure time that define macros like `KOKKOS_ENABLE_CUDA`).  Applications using Approach 2 would need to rebuild, not just relink, if they want to change whether or not to enable CUDA.

**Comment by flying-tiger** [comment]
> you can't just build two separate libraries; you would also need to install two separate sets of header files... need to rebuild, not just relink, if they want to change whether or not to enable CUDA

Yes, sorry. I'm using "link" a bit loosely, in the sense of CMake's "target_link_libraries". You're right; this would mean two sets of generated headers, and if you wanted to switch the application mode, you would need to recompile. But in practice, I think most users of this would just build both ways, e.g.: 

    add_library(my_app_core  ...)

    add_executable(my_app_device ...)
    target_link_libraries(my_app_device PRIVATE my_app_core Kokkos::kokkos)

    add_executable(my_app_host ...)
    target_link_library(my_app_host PRIVATE my_app_core Kokkos::kokkos_host_only)

Then:

    mpiexen -np 1000 my_app_host : -np 100 my_app_device

**Comment by mhoemmen** [comment]
@flying-tiger Thanks for explaining!  This is for the Kokkos team to decide, but it still helps to discuss possible solutions here.

**Comment by jrobcary** [comment]
I'd like to endorse the feature request.  Our use case is that we build an application with CUDA enabled, and we would like to distribute that application for running on hardware lacking a CUDA GPU.  So for us the initialization would skip the CUDA backend if there is no GPU or no GPU assigned to that process.  At this point in time, our Trilinos-using code gracelessly crashes inside of the CUDA initialization.

We could build two versions of the code, just as we build for different OSs, but this does not scale as one considers more backends, eg, HIP, Intel GPUs, etc.  The lack of this feature is basically preventing general binary distribution, which is a killer in the commercial world.

We would prefer Approach 1.  We would look for resources to help with this feature.

**Comment by masterleinad** [comment]
This should be possible already, see https://github.com/kokkos/kokkos/blob/develop/core/unit_test/TestLegionInitialization.cpp. Note that the relevant initialization functions are in the `Impl` namespace and might thus change.

**Comment by jrobcary** [comment]
Very nice and thanks!
We are actually using Tpetra::ScopeGuard(argc, argv), which I understand flows down to Kokkos::initialize.  Is there an example of working through Tpetra::ScopeGuard?
Can one, eg, call Kokkos::initialize as shown, then Tpetra::ScopeGuard, with the second call seeing that Kokkos is already initialized and so not doing it again?

**Comment by masterleinad** [comment]
> so then I just initialize Kokkos, then call Tpetra::ScopeGuard, it seems.

Yes, I would think that should work but I haven't tried any of that.

**Comment by maliberty** [comment]
I'm new to Kokkos but excited.  I have the same concern as @jrobcary about a single binary but I'm unable to follow the description of TestLegionInitialization.cpp solves the problem.   @masterleinad would you elaborate on how initialization should work assuming you can detect whether you want CUDA or OpenMP at runtime startup?

**Comment by masterleinad** [comment]
> would you elaborate on how initialization should work assuming you can detect whether you want CUDA or OpenMP at runtime startup?

You would call
```C++
Kokkos::InitializationSettings kokkos_init_settings;
Kokkos::Impl::pre_initialize(kokkos_init_settings);

if (use_cuda)
  Kokkos::Cuda::impl_initialize(kokkos_init_settings);
if (use_openmp)
  Kokkos::OpenMP::impl_initialize(kokkos_init_settings);

Kokkos::Impl::post_initialize(kokkos_init_settings);
```
instead of `Kokkos::initialize()` and
```C++
Kokkos::Impl::pre_finalize();

if (use_cuda)
  Kokkos::Cuda::impl_finalize();
if (use_openmp)
  Kokkos::OpenMP::impl_finalize();
 
Kokkos::Impl::post_finalize();
```
instead of `Kokkos::finalize()`;

**Comment by jrobcary** [comment]
Given that the above does not call, eg, `Kokkos::initialize(argc, argv)`, is it assuming no MPI?

**Comment by masterleinad** [comment]
`Kokkos::initialize` doesn't have much to do with MPI apart from choosing the device id if there are multiple. The above example is equivalent to calling `Kokkos::initialize`.

**Comment by jrobcary** [comment]
When I try the above, I get

**Comment by jrobcary** [comment]
Trying the above, slightly modified,

```
  Kokkos::InitializationSettings kokkos_init_settings;
  Kokkos::Impl::pre_initialize(kokkos_init_settings);
// Initializing for multiple backends
// https://github.com/kokkos/kokkos/issues/2623#issuecomment-2207543939
#ifdef TXIMPLICIT_HAVE_CUDA
  Kokkos::Cuda::impl_initialize(kokkos_init_settings);
#else
  Kokkos::OpenMP::impl_initialize(kokkos_init_settings);
#endif
  Kokkos::Impl::post_initialize(kokkos_init_settings);
```

(TXIMPLICIT_HAVE_CUDA is true if CUDA is detected.)

I get

```
/home/research/cary/projects/xsimall-dev4/tximplicit/txmbase/tests/kokkosInitTimer.cxx(31): error: name followed by "::" must be a class or namespace name
    Kokkos::OpenMP::impl_initialize(kokkos_init_settings);
            ^
```

So then I tried

```
#ifdef TXIMPLICIT_HAVE_CUDA
  Kokkos::Cuda::impl_initialize(kokkos_init_settings);
#else
  Kokkos::Serial::impl_initialize(kokkos_init_settings);
#endif
```

and got

```
cary@vcloud/.../tests$ ./kokkosInitTimer 
Kokkos::Cuda::initialize WARNING: running kernels compiled for compute capability 7.0 on device with compute capability 8.6 , this will likely reduce potential performance.
terminate called after throwing an instance of 'std::runtime_error'
  what():  Cuda::initialize ERROR : HostSpace::execution_space is not initialized
Aborted (core dumped)
```

So then I tried

```
#ifdef TXIMPLICIT_HAVE_CUDA
  Kokkos::Cuda::impl_initialize(kokkos_init_settings);
#endif
  Kokkos::Serial::impl_initialize(kokkos_init_settings);
```

and got the same error.

????

**Comment by masterleinad** [comment]
You always have to initialize a host execution space first.

**Comment by jrobcary** [comment]
Thanks!  Got past Kokkos initialization.

For Tpetra, is there a similar way of creating the ScopeGuard without passing command-line args?  That is very convenient, as by the time I want to create this, I am deep into the code, having determined whether the particular process was assigned a GPU.
